storage/src/main/scala/coop/rchain/storage/regex/Fsm.scala

```diff
@@ -0,0 +1,811 @@
+    * string of symbols. <https://en.wikipedia.org/wiki/Brzozowski_derivative>
+    * If any of the symbols are not members of the alphabet, that's a NoSuchElementException.
+    * If you fall into oblivion, then the derivative is an FSM accepting no
+    * strings.
+    */
+  def Derive(input: String): Fsm = {
+    var currentState = initialState
+    for (currentSymbol <- input) {
+      val sym = if (alphabet.contains(currentSymbol)) {
+        currentSymbol
+      } else if (HasAnythingElse) {
+        Fsm.AnythingElse
+      } else {
+        throw new NoSuchElementException(
+          "Symbol '" + currentSymbol + "' is not in the source alphabet.")
+      }
+
+      transitions.get(currentState) match {
+        case None => return Fsm.NullFsm(alphabet)
+        case Some(charToStateMap) => {
+          charToStateMap.get(sym) match {
+            case Some(nextState) => currentState = nextState
+            case None            => return Fsm.NullFsm(alphabet)
+          }
+        }
+      }
+    }
+    //we've consumed the input, use the new location as the starting point.
+    new Fsm(alphabet, states, currentState, finalStates, transitions)
+  }
+
+  /**
+    * Return a new FSM such that for every string that self accepts (e.g.
+    * "beer", the new FSM accepts the reversed string ("reeb").
+    */
+  def Reversed: Fsm = {
+    //Reversed FSM uses the same alphabet
+    val resAlphabet = alphabet
+    //Start from a composite "state-set" consisting of all final states.
+    //If there are no final states, this set is empty and we'll find that
+    //no other states get generated.
+    val resInitials = finalStates
+    //Find every possible way to reach the current state-set
+    //using this symbol.
+    def follow(currentStates: Set[Int], currentSymbol: Char): Option[Set[Int]] = {
+      val nextStates = (for {
+        (transitionState, transitionMap) <- transitions
+        currentState <- currentStates
+        if transitionMap.get(currentSymbol).contains(currentState)
+      } yield transitionState).toSet
+
+      if (nextStates.isEmpty) {
+        None
+      } else {
+        Some(nextStates)
+      }
+    }
+    //A state-set is final if the initial state is in it.
+    def isFinal(statesSet: Set[Int]): Boolean = statesSet.contains(initialState)
+    //run crawl, and do not reduce() the result, since reduce() calls us in turn
+    Fsm.crawl(resAlphabet, resInitials, isFinal, follow)
+  }
+
+  /**
+    * Return a finite state machine which will accept any string NOT
+    * accepted by self, and will not accept any string accepted by self.
+    * This is more complicated if there are missing transitions, because the
+    * missing "dead" state must now be reified.
+    */
+  def EverythingBut: Fsm = {
+    def resInitial = initialState :: Nil
+
+    def follow(currentState: List[Int], currentSymbol: Char): Option[List[Int]] = {
+      var next = List[Int]()
+      if (currentState.nonEmpty) {
```

Comment:
You could use `map` or `foreach` instead.

Reply:
Replaced